appbuilder/project/project-base.ts

```diff
@@ -40,5 +61,82 @@ export class Project implements Project.IProjectBase {
 	}
 
 	public startPackageActivity = startPackageActivityNames[TARGET_FRAMEWORK_IDENTIFIERS.Cordova.toLowerCase()];
+
+	public get hasBuildConfigurations(): boolean {
+		return this._hasBuildConfigurations;
+	}
+
+	public get projectInformation(): Project.IProjectInformation {
+		return {
+			projectData: this.projectData,
+			configurationSpecificData: this.configurationSpecificData,
+			hasBuildConfigurations: this.hasBuildConfigurations,
+			configurations: Object.keys(this.configurationSpecificData)
+		};
+	}
+
+	protected abstract validate(): void;
+	protected abstract saveProjectIfNeeded(): void;
+
+	protected readProjectData(): IFuture<void> {
+		return (() => {
+			let projectDir = this.getProjectDir().wait();
+			this.setShouldSaveProject(false);
+			if(projectDir) {
+				let projectFilePath = path.join(projectDir, this.$projectConstants.PROJECT_FILE);
+				try {
+					let data = this.$fs.readJson(projectFilePath).wait();
```

Comment:
this method is waaay too long, can you extract the first part to a method, for example `readProjectFile`:
```TypeScript
let projectFilePath = path.join(projectDir, this.$projectConstants.PROJECT_FILE);
try {
	let { data, shouldSaveProject } = this.readProjectFile(projectFilePath);
	this.projectData = data;
	this.setShouldSaveProject(shouldSaveProject);

	this.validate();
....
}
```